psalm.xml.dist

```diff
@@ -53,5 +41,17 @@
                 <referencedMethod name="Doctrine\DBAL\Connection::getEventManager"/>
             </errorLevel>
         </DeprecatedMethod>
+        <UndefinedClass>
+            <errorLevel type="suppress">
+                <referencedClass name="Doctrine\Common\Cache\Cache"/>
+                <referencedClass name="Doctrine\Common\Cache\CacheProvider"/>
+                <referencedClass name="Doctrine\Common\Cache\Psr6\CacheAdapter"/>
+            </errorLevel>
+        </UndefinedClass>
+        <UndefinedDocblockClass>
+            <errorLevel type="suppress">
+                <referencedClass name="Doctrine\Common\Cache\Cache"/>
+            </errorLevel>
+        </UndefinedDocblockClass>
```

Comment:
These should go away before merging: the fact that we reference these classes, yet they are not explicit dependencies, is a problem.

Reply:
those class referencesses are in `instanceof`-operator to check package availability (before replace cache with bundled `CacheAdapter`) and special `::class` consts in `InvalidArgumentException::fromUnsupportedCache()`. So as for me its safe for optional dependency.

Bytheway, until [doctrine/orm](https://github.com/doctrine/orm/blob/642a20bcff440234afe50c4a2f91c786298d5947/composer.json#L28) will not bump `doctrine/dbal:^4.0` all project will have `doctrine/cache`, but whats the reason to declare it as that package dependency.

As less error-prone approach I could inline suppression:
```
@psalm-suppress UndefinedClass, UnusedPsalmSuppress
```

Reply:
If we say that we go with `doctrine/cache:^2`, then we should most certainly not have any undefined classes at all.

If the support for `doctrine/cache` requires referencing classes from `doctrine/cache`, then the dependency should be declared in `composer.json`.

This applies to `instanceof` and `::class` references alike, but I'll check the patch in more detail when I have more time :)

Reply:
no its not about `doctrine/cache:^2` anymore. We allow v1 - https://github.com/Roave/psr-container-doctrine/pull/72#discussion_r1098606375 - but drop logic specific for this version (i mean instantiation support in `CacheFactory`). But stays logic for both versions:
- wrap into PSR-6 interface adapter required by ORM
- set namespace from `$config['doctrine']['cache'][$configKey]['namespace']` in case when no conventional cache service id [`"doctrine.cache.{$configKey}"`](https://github.com/Roave/psr-container-doctrine/blob/9cfad0cdd159ac86cd1d5d05759fb0007210b05d/src/AbstractFactory.php#L105) (assume its fully build for usage with adapter).

With knowledge that `doctrine/cache:^2` is an intermediate solution for migrating to the PSR-6 cache interface. It would be great to make package optional dependency.

Reply:
`doctrine/cache` should be moved from `require` to `require-dev` then, to satisfy psalm, and not impose it on users.

Reply:
yes, this will solve problem with ugly psalm suppression. It looks like ORM cache types are stable (metadata, query, result, hydration) so no problem that contributors may not aware about disappearing `doctrine/cache` package.

But there is other blocker for optional dependency for newbies - [composer-require-checker](https://github.com/Roave/psr-container-doctrine/blob/607b0ae0f8012eb91a598989959be05f1304437c/.github/workflows/composer-require-checker.yml#L27) - which require add `composer-require-checker.json` with list optional class lists:
```
{
  "symbol-whitelist" : [
    "Doctrine\\Common\\Cache\\Cache",
    "Doctrine\\Common\\Cache\\CacheProvider",
    "Doctrine\\Common\\Cache\\Psr6\\CacheAdapter"
  ]
}
```

Reply:
looks like `composer-require-checker` struggle against complexity caused by optional dependencies, so I will return `doctrine/cache` to `required` dependencies